gma/src/android/native_ad_internal_android.cc

```diff
@@ -138,6 +138,131 @@ Future<AdResult> NativeAdInternalAndroid::LoadAd(const char* ad_unit_id,
+}
+
+Future<void> NativeAdInternalAndroid::PerformClick(const Variant& click_data) {
+  firebase::MutexLock lock(mutex_);
+
+  if (!initialized_) {
+    return CreateAndCompleteFuture(kNativeAdFnPerformClick,
+                                   kAdErrorCodeUninitialized,
+                                   kAdUninitializedErrorMessage, &future_data_);
+  }
+
+  jobject click_bundle = variantmap_to_bundle(click_data);
+
+  if (click_bundle == nullptr) {
+    return CreateAndCompleteFuture(
+        kNativeAdFnPerformClick, kAdErrorCodeInvalidArgument,
+        kUnsupportedVariantTypeErrorMessage, &future_data_);
+  }
+
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  FutureCallbackData<void>* callback_data =
+      CreateVoidFutureCallbackData(kNativeAdFnPerformClick, &future_data_);
+  Future<void> future =
+      MakeFuture(&future_data_.future_impl, callback_data->future_handle);
+  env->CallVoidMethod(
+      helper_, native_ad_helper::GetMethodId(native_ad_helper::kPerformClick),
+      reinterpret_cast<jlong>(callback_data), click_bundle);
+
+  util::CheckAndClearJniExceptions(env);
+
+  return future;
+}
+
+jobject NativeAdInternalAndroid::variantmap_to_bundle(
+    const Variant& variant_data) {
+  if (!variant_data.is_map()) {
+    return nullptr;
+  }
+
+  JNIEnv* env = ::firebase::gma::GetJNI();
+  FIREBASE_ASSERT(env);
+
+  jobject variant_bundle =
+      env->NewObject(util::bundle::GetClass(),
+                     util::bundle::GetMethodId(util::bundle::kConstructor));
+  FIREBASE_ASSERT(variant_bundle);
+
+  for (const auto& kvp : variant_data.map()) {
+    const Variant& key = kvp.first;
+    const Variant& value = kvp.second;
+
+    if (!key.is_string()) {
+      return nullptr;
+    }
+    jstring key_str = env->NewStringUTF(key.string_value());
+
+    if (value.is_int64()) {
+      jlong val_long = (jlong)value.int64_value();
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutLong),
+                          key_str, val_long);
+    } else if (value.is_double()) {
+      jfloat val_float = (jfloat)value.double_value();
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutFloat),
+                          key_str, val_float);
+    } else if (value.is_string()) {
+      jstring val_str = env->NewStringUTF(value.string_value());
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutString),
+                          key_str, val_str);
+      env->DeleteLocalRef(val_str);
+    } else if (value.is_map()) {
+      jobject val_bundle = variantmap_to_bundle(value);
+      env->CallVoidMethod(variant_bundle,
+                          util::bundle::GetMethodId(util::bundle::kPutBundle),
+                          key_str, val_bundle);
+      env->DeleteLocalRef(val_bundle);
+    } else {
+      // Unsupported value type.
```

Comment:
You need to remove the ref to the key_str here, too.

Reply:
Good catch! Addressed in the latest commit.